### DISCUSSION TITLE
References for Code Notes

### DISCUSSION THREAD

**Discussion by SporyTike** [RetroAchievements/RAIntegration#993] [Ideas]
As a developer it sometimes happens that the structure of some memory is the same in multiple locations, for example character slots in an RPG. The code note for one of those slots could look like this:

> Character Slot - Cloud [8 Bytes]
> +00 | Character ID [16bit]
> +02 | Current HP [16bit]
> +04 | Max HP [16bit]
> +06 | Current Armor ID [8bit]
> +07 | Current Weapon ID [8bit]

If you have multiple slots following the same structure you have two ways to approach that:

1. Copy-Paste the structure to every slot. Problem: If you add information or need to make a change all notes need to be updated which is tedious.
2. Take a specific slot and in every other slot you reference to the original slot for more information via comment. Problem: If someone investigates the logic, they need to switch between addresses to understand the logic.

It would be nice to have a way to do inbuild references. While I can see it being hard to implement in the code note input field itself as well as confusing, I think it would be great if at least the address code note in the Asset Editor (and eventually the site) could show such information.

Here some images to visualize my idea:

**Shown Code Note in the Asset Editor**
![image](https://user-images.githubusercontent.com/45672834/230029121-2f193c67-24db-4b4a-ac39-6e029c95d423.png)

**Actual Code Note of address 0x200040 referencing to address 0x000000**
![image](https://user-images.githubusercontent.com/45672834/230028763-b3442151-c353-46b7-9570-e31d24de95d6.png)

**Code Note of address 0x000000, this is the one shown in the Asset Editors ignoring the comment**
![image](https://user-images.githubusercontent.com/45672834/230029474-167490b3-edfa-4f93-bba9-6f72852bc0d3.png)

The way it should work would be to simply replace the @0x000000 reference with the data from that address, so that you can also use it inbetween text like this in case at some point the data slightly changes or you want to reference something like character or weapon IDs.

So this:
> Character Slot - Cloud [10 Bytes]
> @0x000000
> +08 | Wins in a Row [16bit]

Would result into:
> Character Slot - Cloud [10 Bytes]
> +00 | Character ID [16bit]
> +02 | Current HP [16bit]
> +04 | Max HP [16bit]
> +06 | Current Armor ID [8bit]
> +07 | Current Weapon ID [8bit]
> +08 | Wins in a Row [16bit]

In the example I just use address 0x000000 as reference location, not the best way to approach this since it would be better to have such reference information excluded from actual code notes but it's at least doable without changing much of how notes work and there usually is always memory that's unnecessary for the achievement dev.

To avoid infinite loops and very complex reference chains I think a simply 1-layer reference limit would be fine.

**Comment by redwizard42** [comment]
Just some way to define structure notes would just be nice in general as set-wide meta data.

[Structure:CharacterSlot:128Bytes]
+offset blah blah

[Structure:BattleSlot:64Bytes]
+offset1 blah
+offset2 hard

And then you can just tag something as [CharacterSlot]  or [BattleSlot] on an address and it'd ghost out the whole range plus solid highlight the offsets would be nice.  I'd love to see things like this show up proper-like on the website too.

**Comment by Jamiras** [comment]
I was considering something along these lines for annotating an array of structures. It could potentially be used for a single structure by making the array length 1, but you'd probably just note the individual fields at that point.
```
[8x40 bytes] Party Data
+00 | Character ID [16bit]
+02 | Current HP [16bit]
+04 | Max HP [16bit]
+06 | Current Armor ID [8bit]
+07 | Current Weapon ID [8bit]
```
The "40 bytes" is the size of the structure, and the "8x" indicates an array of 8 elements. The entire region would be 320 bytes long, and the structure notes would be repeated every 40 bytes.

If that note was at address $10000, then address $10002 would have an indirect note of:
```
0x10002 (indirect)
Party Data [index 0]
Current HP [16-bit]
```
And address $10056 ($10000 + 40*2 + 6) would have an indirect note of:
```
0x10056 (indirect)
Party Data [index 2]
Current Armor ID [8bit]
```

**Reply by redwizard42** [reply]
Oh, that's a nice annotation for arrays of a structure. I like that.  Would there be a way for the logic definition on the site to also pull in the indirect structure notes (possible as a toggle), even if it's just when hovering over the address?
